core/src/lib.rs

```diff
@@ -300,37 +343,10 @@ impl Schema {
 
     /// Returns an symbolic representation of the schema suitable for applying across Mentat stores.
     pub fn to_edn_value(&self) -> edn::Value {
-        let mut all_values = Vec::new();
-        for (entid, attribute) in &self.schema_map {
-            let mut attribute_map: BTreeMap<edn::Value, edn::Value> = BTreeMap::default();
-            if let Some(ident) = self.get_ident(*entid) {
-                attribute_map.insert(values::DB_IDENT.clone(), edn::Value::NamespacedKeyword(ident.clone()));
-            }
-
-            let value_type = format!("{:?}", attribute.value_type);
-            attribute_map.insert(values::DB_VALUE_TYPE.clone(), edn::Value::NamespacedKeyword(NamespacedKeyword::new("db.type", &value_type.to_lowercase())));
-
-            attribute_map.insert(values::DB_CARDINALITY.clone(), if attribute.multival { values::DB_CARDINALITY_MANY.clone() } else { values::DB_CARDINALITY_ONE.clone() });
-
-            if attribute.unique == Some(attribute::Unique::Value) {
-                attribute_map.insert(values::DB_UNIQUE.clone(), values::DB_UNIQUE_IDENTITY.clone());
-            }
-            
-            if attribute.index {
-                attribute_map.insert(values::DB_INDEX.clone(), edn::Value::Boolean(true));
-            }
-
-            if attribute.fulltext {
-                attribute_map.insert(values::DB_FULLTEXT.clone(), edn::Value::Boolean(true));
-            }
-
-            if attribute.component {
-                attribute_map.insert(values::DB_IS_COMPONENT.clone(), edn::Value::Boolean(true));
-            }
-
-            all_values.push(edn::Value::Map(attribute_map));
-        }
-        edn::Value::Vector(all_values)
+        edn::Value::Vector((&self.schema_map).into_iter()
```

Comment:
Yeah, it's tricky.

The reason for doing it this way is that it's making it very clear what you're doing: turning a thing into an iterator, walking it in order (the only thing you can do with an iterator!), and collecting the results into a vector. It might initially seem less readable than a loop and a mutable collection — that's closures for ya — but it imposes obvious compiler-enforced constraints on the program that _later_ mean you can read it with much more confidence.

Having a `let mut foo = Vec::new()` _implies_ to a Rusty reader several things:

- You don't know how big the output will be, otherwise you'd use `with_capacity`. (This is a sidenote, 'cos of course you could use `w_c` here.)
- You're not doing something that could be expressed with `collect`, otherwise you would be using `collect` — e.g., you're adding two things in some steps of the iteration.
- You might be passing that mutable reference somewhere — manually implementing your own flat map, say, or collecting into the same vector from iterations over two things.

The `collect` form is also more efficient: `collect` can ask the iterator for a hint (see `size_hint`), and an iterator over a fixed collection knows its size, and can pass it through a `Map` iterator. You'd need to manually remember to use `with_capacity` to get the same effect. Rust closures are very cheap, usually eliminated entirely by the compiler, so there's no perf worry here.

In a broader context, you'll find that the iterator/map form is amenable to switching to parallel computation with a single-word change; a `for` loop not so much, precisely because of the constraints the iterator approach opts into. Fun stuff!